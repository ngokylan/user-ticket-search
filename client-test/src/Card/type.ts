import { ReactNode } from 'react';
import { BadgeProps } from '../Badge';

export interface CardProps {
  /** Card id attribute, optional id */
  id?: string;
  /** Card heading title */
  heading?: ReactNode;
  /** Card description */
  description?: string;
  /** Card content */
  children: ReactNode;
  /** Additional class */
  className?: string;
  /** Make height 100% */
  isFullHeight?: boolean;
  /** To place option content to the right side of heading. such as Button */
  option?: ReactNode;
  /** small heading above heading */
  prefix?: string;
  /** image in the header */
  image?: string;
  /** array of the badge components which are displayed on image */
  badges?: BadgeProps[];
  /** function that is triggered if the card is clicked */
  onClick?: Function;
  /** Role is for WCAG accessibility compliance (optional) */
  role?: string;
}
