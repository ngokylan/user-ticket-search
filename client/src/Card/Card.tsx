import React, { Component } from 'react';
import { CardProps } from './type';
import { Heading, Paragraph, Text } from '../Typography';
import Image from '../Image';
import Badge from '../Badge';
import { getClass } from '../_lib/helper';
import './Card.scss';

class Card extends Component<CardProps> {
  renderBadges() {
    const { id, badges } = this.props;
    const keyTemplate = 'card-badge-' + id + '-';
    return (
      badges &&
      badges.map((badge, index) => {
        return (
          <Badge
            icon={badge.icon}
            type={badge.type}
            role={badge.role}
            key={keyTemplate + index}
            data-test-id={keyTemplate + index}
          >
            {badge.children}
          </Badge>
        );
      })
    );
  }

  handleClick = () => {
    const { onClick } = this.props;
    if (onClick) {
      onClick();
    }
    return false;
  };
  handleKeyPress = (e: React.KeyboardEvent) => {
    if (e.key === 'Enter') {
      this.handleClick();
    }
  };

  render() {
    const {
      id,
      heading,
      children,
      description,
      className,
      isFullHeight,
      option,
      prefix,
      image,
      badges,
      onClick,
      role,
    } = this.props;

    return (
      <div
        role={role}
        id={id}
        data-testid={`card-${id || 'default'}`}
        className={getClass('card', className, {
          'is-full-height': isFullHeight,
          'is-clickable': onClick,
        })}
        onClick={onClick ? this.handleClick : undefined}
        onKeyPress={this.handleKeyPress}
        tabIndex={onClick ? 0 : undefined}
      >
        {(image || badges) && (
          <div className="card__header-image">
            {image && (
              <Image
                src={image}
                width="100%"
                height="100%"
                alt={typeof heading === 'string' ? heading : 'Course image'}
              />
            )}
            <div className="card__header-badges">
              {this.renderBadges()}
            </div>
          </div>
        )}

        {option && <div className="card__option">{option}</div>}
        {(prefix || heading || description) && (
          <div className="card__header">
            {heading && (
              <Heading className="card__heading">
                {prefix && (
                  <Text
                    size="sm"
                    color="gray"
                    htmlTag="small"
                    className="d-block mb-2 card__heading-prefix"
                  >
                    {prefix}
                  </Text>
                )}
                {heading}
              </Heading>
            )}
            {description && <Paragraph>{description}</Paragraph>}
          </div>
        )}
        <div className="card__content">
          <div className="card__content-inner">{children}</div>
        </div>
      </div>
    );
  }
}

export default Card;
