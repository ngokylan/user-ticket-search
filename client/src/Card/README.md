Use tag <Card /> to create a card/panel. 

### When To Use

### Examples

Default card
```js
<Card>Content</Card>
```

Card with heading
```js
<Card heading="Heading">Content</Card>
```

Card with description
```js
<Card description="Description">
  Content
</Card>
```

Card with long description
```js
  <Card heading="Card title">
    Lorem ipsum dolor sit amet, consectetur adipisicing elit. Asperiores assumenda beatae, cumque dicta
    dolor doloremque eius excepturi facilis fugit iure laudantium minima nam, nostrum perspiciatis
    quidem similique, sit tempora voluptatibus?
  </Card>
```  