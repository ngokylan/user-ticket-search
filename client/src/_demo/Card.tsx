import React, { Component } from 'react';

import { Heading, Text, Paragraph } from '../Typography';
import Card from '../Card';
import Badge from '../Badge';
import { Row, Col } from '../Grid';

class CardDemo extends Component {
  render() {   
    return (
      <>
        <Heading className="mt-4 mb-3">Course Card</Heading>
        <Row>
          <Col xs={12} md={8} className="mb-3">
            <Card
              isFullHeight={true}
              prefix="Policies and Procedures"
              heading="Figma Tutorial: Device Frames and Scrolling"                            
              className=""
              onClick={() => {
                console.log('click card!');
              }}
            >
              <Badge type="warning" className="mb-3">
                Not Started
              </Badge>
              <Row isNoGutters={true}>
                <Col isAuto={true}>
                  <Text size="sm" color="gray">
                    Completion
                  </Text>
                  <Row isNoGutters={true} wrap="wrap" alignItems="center">
                    <Col isAuto={true} grow={0}>
                      <Paragraph>Jul 30,2019</Paragraph>
                    </Col>
                    <Col isAuto={true} className="ml-2" />
                  </Row>
                </Col>
                <Col
                  isAuto={true}
                  className="d-flex justify-content-end pr-0 -mr-3"
                >                
                </Col>
              </Row>
            </Card>
          </Col>        
        </Row>
      </>
    );
  }
}

export default CardDemo;
