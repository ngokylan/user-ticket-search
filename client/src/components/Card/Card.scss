@import '../../_lib/style/index';

$ComponentClassName: '.card';

#{$ComponentClassName} {
  box-shadow: $shadow-100;
  border-radius: 6px;
  padding: 24px;
  background-color: #fff;
  margin-bottom: 24px;
  outline: 0;

  &__header-image {
    position: relative;
    overflow: hidden;
    margin-top: -24px;
    margin-left: -24px;
    margin-right: -24px;
    border-radius: 3px 3px 0 0;
    margin-bottom: 14px;
    min-width: calc(100% + 48px);
    height: 116px;

    .image {
      object-fit: cover;
    }

    #{$ComponentClassName}__header-badges {
      position: absolute;
      bottom: 4px;
      right: 4px;

      .badge {
        margin-left: 4px;
      }
    }

    & + #{$ComponentClassName}__header {
      margin-bottom: 16px;
    }
  }

  &__header {
    margin-bottom: 48px;
    .card__heading {
      overflow-wrap: break-word;
    }
  }

  &__option {
    float: right;
    margin-top: -8px;
    margin-right: -8px;
  }

  &--is-full-height {
    min-height: 100%;
    display: flex;
    flex-direction: column;
    margin-bottom: 0; // to remove unexpected height

    #{$ComponentClassName}__header-image
      + #{$ComponentClassName}__header
      + #{$ComponentClassName}__content {
      flex-grow: 1;
      display: flex;
      flex-wrap: wrap;
      align-items: flex-end;

      #{$ComponentClassName}__content-inner {
        min-width: 100%;

        > .divider {
          margin-left: -24px;
          margin-right: -24px;
        }
      }
    }
  }

  &--is-clickable {
    cursor: pointer;
    position: relative;
    transition: $transition-speed-normal;
    &:before {
      transition: $transition-speed-normal;

      content: '';
      display: block;
      position: absolute;
      width: calc(100% + 4px);
      height: calc(100% + 4px);
      left: -2px;
      top: -2px;
      border: 2px solid $primary-light;
      border-radius: 5px;
      opacity: 0;
    }
    &:hover {
      box-shadow: $shadow-300;
    }
    &:focus {
      outline: none;
      &:before {
        opacity: 1;
      }
    }
    &:active {
      background-color: $gray-rgba-300;
    }
  }
}

.row {
  > #{$ComponentClassName} {
    width: 100%;
    max-width: 368px;
    margin-left: 12px;
    margin-right: 12px;
  }
}
