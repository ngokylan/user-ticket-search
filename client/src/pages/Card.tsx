import React, { Component } from 'react';
import { Text, Paragraph } from '../components/Typography';
import Card from '../components/Card';
import Badge from '../components/Badge';
import { Row, Col } from '../components/Grid';

class CardDemo extends Component {
  render() {   
    return (
      <>        
        <Row>
          <Col xs={12} md={8} className="mb-3">
            <Card
              isFullHeight={true}
              prefix="Policies and Procedures"
              heading="Figma Tutorial: Device Frames and Scrolling"                            
              className=""
              onClick={() => {
                console.log('click card!');
              }}
            >
              <Badge type="warning" className="mb-3">
                Not Started
              </Badge>
              <Row isNoGutters={true}>
                <Col isAuto={true}>
                  <Text size="sm" color="gray">
                    Completion
                  </Text>
                  <Row isNoGutters={true} wrap="wrap" alignItems="center">
                    <Col isAuto={true} grow={0}>
                      <Paragraph>Jul 30,2019</Paragraph>
                    </Col>
                    <Col isAuto={true} className="ml-2" />
                  </Row>
                </Col>
                <Col
                  isAuto={true}
                  className="d-flex justify-content-end pr-0 -mr-3"
                >                
                </Col>
              </Row>
            </Card>
          </Col>        
        </Row>
      </>
    );
  }
}

export default CardDemo;
