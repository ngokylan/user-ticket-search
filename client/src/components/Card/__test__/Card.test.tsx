import React from 'react';
import { mount, shallow } from 'enzyme';
import Card, { CardProps } from '..';

const defaultProps: CardProps = undefined;
const mountComponent = (customProps: any = {}) => {
  const props = { ...defaultProps, ...customProps };
  const wrapper = mount(<Card {...props} />);
  const component = wrapper.find('.card');
  const instance = wrapper.instance();
  return { props, wrapper, component, instance };
};

describe('Card', () => {
  describe('rendering', () => {
    it('should render component without any crash', () => {
      const { wrapper } = mountComponent();
      expect(wrapper).toHaveLength(1);
    });

    it.each([
      ['Heading exist', { heading: 'Title' }],
      ['Description exist', { description: 'Description' }],
    ])('should be match with snapshot %s', (name, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper).toMatchSnapshot();
    });

    it.each([['role', 'test'], ['role', undefined]])(
      'should be match with snapshot when props %s is %s',
      (propsName, propsValue) => {
        const { wrapper } = mountComponent({ [propsName]: propsValue });
        expect(wrapper).toMatchSnapshot();
      }
    );
  });

  describe('props', () => {
    describe('default', () => {
      it('should have expected default props', () => {
        expect(JSON.stringify(defaultProps)).toBe(
          JSON.stringify(Card.defaultProps)
        );
      });

      it.each([
        ['card-default', {}],
        ['card-custom-id', { id: 'custom-id' }],
      ])('should render test ID properly', (expectedTestId, props) => {
        const { component } = mountComponent(props);
        expect(component.prop('data-testid')).toBe(expectedTestId);
      });
    });

    describe('children', () => {
      it('should render children', () => {
        const { wrapper } = mountComponent({ children: 'Hello world' });
        expect(wrapper.find('.card__content').text()).toBe('Hello world');
      });
    });

    describe('custom', () => {
      it.each([['id', 'test', { id: 'test' }]])(
        'should render prop %s as %s',
        (name, expected, props) => {
          const { component } = mountComponent(props);
          expect(component.prop(name)).toBe(expected);
        }
      );

      it.each([['role', 'test', { role: 'test' }]])(
        'should render prop %s as %s',
        (name, expected, props) => {
          const { component } = mountComponent(props);
          expect(component.prop(name)).toBe(expected);
        }
      );
    });
  });

  describe('interaction', () => {
    it.each([
      ['should', '.card', 1, {}],
      ['should', '.custom-class', 2, { className: 'custom-class' }],
      ['should', '.card--is-full-height', 1, { isFullHeight: true }],
      ['should not', '.card--is-full-height', 0, { isFullHeight: false }],
      ['should', '.card__option', 1, { option: 'some option' }],
      ['should not', '.card__option', 0, {}],
      ['should', '.card__content', 1, {}],
      ['should', '.card__header', 1, { heading: 'some heading' }],
      ['should', '.card__header', 1, { description: 'some description' }],
      ['should', 'Heading', 1, { heading: 'some heading' }],
      ['should not', 'Heading', 0, {}],
      ['should', 'Paragraph', 1, { description: 'some description' }],
      ['should not', 'Paragraph', 0, {}],
      ['should', 'Text', 1, { heading: 'some heading', prefix: 'some prefix' }],
      ['should not', 'Text', 0, {}],
      ['should', 'Image', 1, { image: 'some image url' }],
      ['should not', 'Image', 0, {}],
      [
        'should',
        'Badge',
        1,
        {
          badges: [{ children: 'some children text' }],
        },
      ],
      ['should not', 'Badge', 0, {}],
    ])('%s have %s', (name, element, expected, props) => {
      const { wrapper } = mountComponent(props);
      expect(wrapper.find(element)).toHaveLength(expected);
    });

    describe('Interaction', () => {
      it('handleClick should call onClick', () => {
        const props = {
          onClick: jest.fn(),
        };
        const { instance } = mountComponent(props);
        instance.handleClick();
        expect(props.onClick).toBeCalled();
      });
      it('handleKeyPress Enter should call onClick', () => {
        const props = {
          onClick: jest.fn(),
        };
        const { instance } = mountComponent(props);
        instance.handleKeyPress({ key: 'Enter' } as React.KeyboardEvent);
        expect(props.onClick).toBeCalled();
      });
    });
  });
});
