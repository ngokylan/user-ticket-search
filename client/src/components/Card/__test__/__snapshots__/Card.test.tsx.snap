// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card rendering should be match with snapshot Description exist 1`] = `
<div
  className="card"
  data-testid="card-default"
  onKeyPress={[Function]}
>
  <div
    className="card__header"
  >
    <p
      className="-typography text-size-rg"
      data-testid="-typography-default"
    >
      Description
    </p>
  </div>
  <div
    className="card__content"
  >
    <div
      className="card__content-inner"
    />
  </div>
</div>
`;

exports[`Card rendering should be match with snapshot Heading exist 1`] = `
<div
  className="card"
  data-testid="card-default"
  onKeyPress={[Function]}
>
  <div
    className="card__header"
  >
    <h3
      className="sc-dlfnbm typography card__heading heading-size-rg"
      data-testid="typography-default"
    >
      Title
    </h3>
  </div>
  <div
    className="card__content"
  >
    <div
      className="card__content-inner"
    />
  </div>
</div>
`;

exports[`Card rendering should be match with snapshot when props role is test 1`] = `
<div
  className="card"
  data-testid="card-default"
  onKeyPress={[Function]}
  role="test"
>
  <div
    className="card__content"
  >
    <div
      className="card__content-inner"
    />
  </div>
</div>
`;

exports[`Card rendering should be match with snapshot when props role is undefined 1`] = `
<div
  className="card"
  data-testid="card-default"
  onKeyPress={[Function]}
>
  <div
    className="card__content"
  >
    <div
      className="card__content-inner"
    />
  </div>
</div>
`;
